services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py

```diff
@@ -1,68 +1,203 @@
 
-    # 2. Remove nodes that are gone
-    await utils_docker.remove_monitored_down_nodes(docker_client, monitored_nodes)
+async def _find_terminateable_nodes(
+    app: FastAPI, monitored_nodes: list[Node]
+) -> list[tuple[Node, EC2InstanceData]]:
+    app_settings: ApplicationSettings = app.state.settings
+    assert app_settings.AUTOSCALING_NODES_MONITORING  # nosec
+    docker_client = get_docker_client(app)
+
+    # NOTE: we want the drained nodes where no monitored service is running anymore
+    drained_empty_nodes = [
+        node
+        for node in monitored_nodes
+        if (
+            await utils_docker.compute_node_used_resources(
+                docker_client,
+                node,
+                service_labels=app_settings.AUTOSCALING_NODES_MONITORING.NODES_MONITORING_SERVICE_LABELS,
+            )
+            == Resources.create_as_empty()
+        )
+        and (node.Spec is not None)
+        and (node.Spec.Availability == Availability.drain)
+    ]
+    assert app_settings.AUTOSCALING_EC2_INSTANCES  # nosec
+    if not drained_empty_nodes:
+        # there is nothing to terminate here
+        return []
 
-    # 3. Scale up nodes if there are pending tasks
-    pending_tasks = await utils_docker.pending_service_tasks_with_insufficient_resources(
-        docker_client,
-        service_labels=app_settings.AUTOSCALING_NODES_MONITORING.NODES_MONITORING_SERVICE_LABELS,
-    )
-    await rabbitmq.post_state_message(
-        app,
-        monitored_nodes,
-        cluster_total_resources,
-        cluster_used_resources,
-        pending_tasks,
+    # get the corresponding ec2 instance data
+    # NOTE: some might be in the process of terminating and will not be found
+    ec2_client = get_ec2_client(app)
+    drained_empty_ec2_instances = await asyncio.gather(
+        *(
+            ec2_client.get_running_instance(
+                app_settings.AUTOSCALING_EC2_INSTANCES,
+                tag_keys=[
+                    "io.simcore.autoscaling.version",
+                ],
+                instance_host_name=node.Description.Hostname,
+            )
+            for node in drained_empty_nodes
+            if node.Description and node.Description.Hostname
+        ),
+        return_exceptions=True,
     )
 
+    terminateable_nodes: list[tuple[Node, EC2InstanceData]] = []
+    for node, ec2_instance_data in zip(
+        drained_empty_nodes, drained_empty_ec2_instances
+    ):
+        if isinstance(ec2_instance_data, Ec2InstanceNotFoundError):
+            # skip if already terminating
+            continue
+        # NOTE: AWS price is hourly based (e.g. same price for a machine used 2 minutes or 1 hour, so we wait until 55 minutes)
+        elapsed_time_since_launched = (
+            datetime.utcnow().replace(tzinfo=timezone.utc)
+            - ec2_instance_data.launch_time
+        )
+        elapsed_time_since_full_hour = elapsed_time_since_launched % timedelta(hours=1)
```

Comment:
Will this work if the time changes to summer time?

Ridiculous question, I know, but let's try to avoid super-weird errors two times per year :D

I guess it should be fine since you explicitly use UTC

Reply:
that is precisely why I use datetime.datetime, such that the time zone is set (default is UTC). And, for the case here what I check is anyway only how many minutes of a full hour elapsed.
from @Surfict I know that using an EC2 instance for 2 minutes or 59 minutes costs the same.
any started hour costs the same as the full hour. therefore I'm only interested in the number of minutes in each hour. So I do think that summer time, whatever time does not really matter here.

I will nevertheless add a test for negative hour difference to be on the same side.

Reply:
```python
datetime.timedelta(hours=-1, minutes=5) % datetime.timedelta(hours=1)
datetime.timedelta(seconds=300)
```

Reply:
so it works